operator/src/main/java/org/keycloak/operator/controllers/WatchedSecretsController.java

```diff
@@ -0,0 +1,155 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+
+import static io.javaoperatorsdk.operator.api.reconciler.Constants.WATCH_CURRENT_NAMESPACE;
+
+@ApplicationScoped
+@ControllerConfiguration(namespaces = WATCH_CURRENT_NAMESPACE, labelSelector = Constants.KEYCLOAK_COMPONENT_LABEL + "=" + WatchedSecrets.WATCHED_SECRETS_LABEL_VALUE)
+public class WatchedSecretsController implements Reconciler<Secret>, EventSourceInitializer<Secret>, WatchedSecrets {
+
+    @Inject
+    KubernetesClient client;
+
+    private final SimpleInboundEventSource eventSource = new SimpleInboundEventSource();
+
+    private volatile IndexerResourceCache<Secret> secrets;
+
+    @Override
+    public Map<String, EventSource> prepareEventSources(EventSourceContext<Secret> context) {
+        this.secrets = context.getPrimaryCache();
+        return Map.of();
+    }
+
+    @Override
+    public UpdateControl<Secret> reconcile(Secret resource, Context<Secret> context) throws Exception {
+        // find all statefulsets to notify
+        //  - this could detect whether the reconciliation is even necessary if we track individual hashes
+        var ret = client.apps().statefulSets().inNamespace(resource.getMetadata().getNamespace())
+                .withLabels(Constants.DEFAULT_LABELS).list().getItems().stream()
+                .filter(statefulSet -> getSecretNames(statefulSet).contains(resource.getMetadata().getName()))
+                .map(statefulSet -> new ResourceID(statefulSet.getMetadata().getName(),
+                        resource.getMetadata().getNamespace()))
+                .collect(Collectors.toSet());
+
+        if (ret.isEmpty()) {
+            Log.infof("Removing label from Secret \"%s\"", resource.getMetadata().getName());
+
+            return UpdateControl.updateResource(new SecretBuilder(resource)
+                    .editMetadata()
+                    .removeFromLabels(Constants.KEYCLOAK_COMPONENT_LABEL)
+                    .endMetadata()
+                    .build());
+        } else {
+            ret.forEach(eventSource::propagateEvent);
+        }
+
+        return UpdateControl.noUpdate();
+    }
+
+    @Override
+    public EventSource getWatchedSecretsEventSource() {
+        return eventSource;
+    }
+
+    @Override
+    public void processWatched(List<String> desiredWatchedSecretsNames, Keycloak keycloakCR, StatefulSet deployment) {
+        List<Secret> currentSecrets = fetchSecrets(desiredWatchedSecretsNames, keycloakCR.getMetadata().getNamespace());
+        deployment.getMetadata().getAnnotations().put(Constants.KEYCLOAK_WATCHING_ANNOTATION, desiredWatchedSecretsNames.stream().collect(Collectors.joining(";")));
+        deployment.getSpec().getTemplate().getMetadata().getAnnotations().put(Constants.KEYCLOAK_WATCHED_SECRET_HASH_ANNOTATION, getSecretHash(currentSecrets));
```

Comment:
> I think we could count with this use case already, but I'm also fine with keeping as is and rewriting the logic for it when we need it. @shawkins WDYT?

Yes that sounds like a future case that would have required changes regardless.

> To do so, I think we could set the Watched Secrets annotation on the CR itself (with some flag if restart is required) instead of setting it on the StS.

Ideally you could do this as inline as possible with the usage of the secret.  For things like SecretKeySelector there's an optional flag already, but it's a stretch to have that also mean don't watch.  You could have keycloak extension to SecretKeySelector (and SecretEnvSource if that's ever used) or other places where a secret is referenced that adds a watching flag.

Reply:
What I meant is there will be use cases where a change in a watched secret will resolve to e.g. a REST API call to Keycloak rather than a deployment restart. Not all secrets will be mounted, probably. But that's something for the future. We might as well just leverage Keycloak Vault and mount the secrets as files, which would required deployment restart on a change.

Let's continue with the current approach then.